File: 2021_04_18_make_a_lisp_in_csharp/Mal/Mal.Tests/ReaderTests.cs

using System;
using Microsoft.VisualStudio.TestTools.UnitTesting;
using static Mal.Types;
using static Mal.Reader;
using static PowerFP.LListM;
using PowerFP;

namespace Mal.Tests
{
    [TestClass]
    public class ReaderTests
    {
        [TestMethod]
        public void ReadAtomTest()
        {
            Assert.AreEqual(TrueV, ReadAtom("true"));
            Assert.AreEqual(FalseV, ReadAtom("false"));
            Assert.AreEqual(NilV, ReadAtom("nil"));
            Assert.AreEqual(new Keyword("hej"), ReadAtom(":hej"));
            Assert.AreEqual(new Symbol("hej", NilV), ReadAtom("hej"));
            Assert.AreEqual(new Str("hej"), ReadAtom("\"hej\""));
            Assert.AreEqual(new Str("hej \\ \n "), ReadAtom("\"hej \\\\ \\n \""));

            Assert.ThrowsException<Exception>(() => ReadAtom("\""));
            Assert.ThrowsException<Exception>(() => ReadAtom("\"hej"));
            Assert.ThrowsException<Exception>(() => ReadAtom("\"hej\" "));
        }


        [TestMethod]
        public void MalsToKeyValuePairsTest()
        {
            Assert.AreEqual(null, MalsToKeyValuePairs(null));
            Assert.AreEqual(
                new (MalType Key, MalType Value)[] { (new Str("a"), TrueV), (new Keyword("b"), FalseV) }.ToLList(),
                MalsToKeyValuePairs(LListFrom<MalType>(new Str("a"), TrueV, new Keyword("b"), FalseV)));

            Assert.ThrowsException<Exception>(() => MalsToKeyValuePairs(LListFrom<MalType>(new Str("a"))));
            Assert.ThrowsException<Exception>(() => MalsToKeyValuePairs(LListFrom<MalType>(TrueV, TrueV)));
        }


        [TestMethod]
        public void ReadListTest()
        {
            Assert.AreEqual(
                new ListReader(new("a", null), LListFrom<MalType>(new Str("a"), FalseV, new Number(12))),
                ReadList(LListFrom("\"a\"", "false", "12", ")", "a"), ")"));

            Assert.ThrowsException<Exception>(() => ReadList(null, ")"));
            Assert.ThrowsException<Exception>(() => ReadList(LListFrom("\"a\"", "false"), ")"));
        }

        [TestMethod]
        public void ReadMetaTest()
        {
            var result = ListFrom(new Symbol("with-meta", NilV), FalseV, TrueV);

            Assert.AreEqual(new FormReader(new("aa", null), result), ReadMeta(LListFrom("true", "false", "aa")));
            Assert.AreEqual(new FormReader(null, result), ReadMeta(LListFrom("true", "false")));

            Assert.AreEqual(new FormReader(null, null), ReadMeta(LListFrom("true")));
            Assert.AreEqual(new FormReader(null, null), ReadMeta(LListFrom<string>()));
        }

        [TestMethod]
        public void ReadMacroTest()
        {
            const string macroToken = "quote";
            var result = ListFrom(new Symbol(macroToken, NilV), TrueV);

            Assert.AreEqual(new FormReader(new("aa", null), result), ReadMacro(LListFrom("true", "aa"), macroToken));
            Assert.AreEqual(new FormReader(null, result), ReadMacro(LListFrom("true"), macroToken));

            Assert.AreEqual(new FormReader(null, null), ReadMacro(LListFrom<string>(), macroToken));
        }


        [TestMethod]
        public void ReadFormTest()
        {
            Assert.AreEqual(new FormReader(null, null), ReadForm(null));
            Assert.AreEqual(
                new FormReader(new("aa", null), ListFrom(new Symbol("with-meta", NilV), FalseV, TrueV)),
                ReadForm(LListFrom("^", "true", "false", "aa")));
            Assert.AreEqual(
                new FormReader(new("aa", null), ListFrom(new Symbol("quote", NilV), TrueV)),
                ReadForm(LListFrom("'", "true", "aa")));
            Assert.AreEqual(
                new FormReader(new("aa", null), ListFrom(TrueV, FalseV)),
                ReadForm(LListFrom("(", "true", "false", ")", "aa")));
            Assert.AreEqual(
                new FormReader(new("aa", null), ListFrom(TrueV, FalseV, ListFrom(NilV))),
                ReadForm(LListFrom("(", "true", "false", "(", "nil", ")", ")", "aa")));

            var (Tokens, Mal) = ReadForm(LListFrom("{", ":name", "\"marcin\"", ":age", "30", "}", "aa"));
            Assert.AreEqual(new("aa", null), Tokens);
            Assert.IsTrue(Mal is Map);
            Assert.IsTrue(Types.MalEqual(
                new Map(new() { { new Keyword("name"), new Str("marcin") }, { new Keyword("age"), new Number(30) } }, NilV),
                Mal!));
        }
    }
}

File: 2021_10_21_make_a_lisp_in_fsharp/MiniMal/MiniMal.Tests/ReaderTests.fs
namespace MiniMal.Tests

open Microsoft.VisualStudio.TestTools.UnitTesting
open Types
open Reader
open TestUtils

[<TestClass>]
type ReaderTests() =

    [<TestMethod>]
    member this.ReadAtomTest() =
        assertMalEquals (readAtom "false") False
        assertMalEquals (readAtom "true") True
        assertMalEquals (readAtom "nil") Nil
        assertMalEquals (readAtom "abc") (Symbol("abc"))
        assertMalEquals (readAtom "\"abc\"") (Str("abc"))
        assertMalEquals (readAtom "123") (Number(123.))
        assertThrowsException (fun () -> readAtom "\"abc")


    [<TestMethod>]
    member this.ReadListTest() =
        let tokens = [ "a"; "123"; ")" ]
        let { ReadListResult.Result = items } = readList tokens ")"
        match items with
        | [ first; second ] ->
            assertMalEquals first (Symbol("a"))
            assertMalEquals second (Number(123.))
        | _ -> Assert.Fail("incorrect list created")


    [<TestMethod>]
    member this.ListToMapTest() =
        let nameProp = Str("name")
        let name = Str("adam")
        let ageProp = Str("age")
        let age = Number(20.)
        assertThrowsException (fun () -> listToMap ([ nameProp ]))
        assertThrowsException (fun () -> listToMap ([ nameProp; name; ageProp ]))
        assertThrowsException (fun () -> listToMap ([ nameProp; name; age; age ]))
        let malMap =
            listToMap ([ nameProp; name; ageProp; age ])
        match malMap with
        | MalMap (map) ->
            Assert.AreEqual(2, Map.count map)
            assertMalEquals name (Map.find "name" map)
            assertMalEquals age (Map.find "age" map)
        | _ -> Assert.Fail("incorrect map created")

    [<TestMethod>]
    member this.ReadFormTest() =
        assertMalEquals (readForm [ "nil"; "true" ]).Result.Value Nil

        assertMalEquals
            (readForm [ "("; "+"; "1"; "2"; ")" ])
                .Result
                .Value
            (MalList([ Symbol("+"); Number(1.); Number(2.) ], List))

        assertMalEquals
            (readForm [ "("
                        "1"
                        "["
                        "2"
                        "]"
                        ")" ])
                .Result
                .Value
            (MalList([ Number(1.); MalList([ Number(2.) ], Vector) ], List))

        assertMalEquals
            (readForm [ "{"
                        "\"name\""
                        "\"adam\""
                        "}" ])
                .Result
                .Value
            (MalMap(Map [ "name", Str("adam") ]))
